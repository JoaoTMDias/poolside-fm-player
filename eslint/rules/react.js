/**
 * React specific rules
 *
 * @author Jo√£o Dias (joao.dias@wit-software.com)
 * @since 1.0.0
 */
module.exports = {
	extends: ['react-app', 'airbnb'],
	plugins: ['react', 'react-hooks'],
	rules: {
		'react/jsx-wrap-multilines': 0,
		'react/jsx-closing-bracket-location': 0,
		'react/jsx-curly-spacing': 0,
		'react/jsx-equals-spacing': 0,
		'react/jsx-first-prop-new-line': 0,
		'react/jsx-indent': 0,
		'react/jsx-indent-props': 0,
		'react/jsx-max-props-per-line': 0,
		'react/jsx-tag-spacing': 0,
		'react/jsx-closing-tag-location': 0,
		'react/jsx-filename-extension': [
			'error',
			{
				extensions: ['.js', '.jsx', '.ts', '.tsx'],
			},
		],
		'react-hooks/rules-of-hooks': 'error',
		'react-hooks/exhaustive-deps': 'warn',
		'react/prop-types': 'off',
		'react/require-default-props': 'off',

		'jsx-a11y/href-no-hash': 'off',
		'jsx-a11y/anchor-is-valid': [
			'warn',
			{
				aspects: ['invalidHref'],
			},
		],
	},
};
