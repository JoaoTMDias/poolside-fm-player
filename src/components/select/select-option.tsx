import React, { useRef } from "react";
import { ISelectOption } from "data/constants";
import { useFocusOnElement, useRovingTabIndex } from "helpers/custom-hooks/roving-index";
import { useDidMount } from "helpers";

interface ISelectOptionProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
	id: string;
	option: ISelectOption;
	disabled?: boolean;
	onSelect: () => void;
	children?: React.ReactNode;
}

interface ISelectInputOptionLabel {
	isNew?: boolean;
	label: string;
}

/**
 * Returns the label for each option
 *
 * @param {ISelectInputOptionLabel} props
 * @returns {JSX.Element}
 */
const SelectInputOptionLabel: React.FunctionComponent<ISelectInputOptionLabel> = ({
	isNew,
	label,
}): JSX.Element => {
	return (
		<span className="ui-label">
			{label}
			{isNew && <span className="is-new">New</span>}
		</span>
	);
};

export const SelectOption: React.FC<ISelectOptionProps> = ({
	id,
	option,
	disabled = false,
	onSelect,
	...rest
}) => {
	const refId = React.useRef<string>(id);
	const ref = useRef<HTMLButtonElement>(null);
	const [tabIndex, focused, handleOnKeyPress, handleClick] = useRovingTabIndex(ref, disabled, id);

	const isSelected = !!(tabIndex === 0);

	useFocusOnElement<HTMLButtonElement>(focused, ref);


	function _onClick(event: React.MouseEvent<HTMLButtonElement, MouseEvent>) {
		event.preventDefault();

		handleClick();
		onSelect();
	}

	return (
		<li role="option" className="select-input__option" aria-selected={isSelected}>
			<button
				ref={ref}
				id={refId.current}
				type="button"
				className="select-input__option__label"
				disabled={disabled}
				onKeyUp={handleOnKeyPress}
				onClick={_onClick}
				tabIndex={tabIndex}
				aria-disabled={disabled}
				{...rest}
			>
				<SelectInputOptionLabel isNew={option.isNew} label={`${option.label}`} />
			</button>
		</li>
	);
};

SelectOption.defaultProps = {
	disabled: false,
	type: "button",
};

export default SelectOption;
