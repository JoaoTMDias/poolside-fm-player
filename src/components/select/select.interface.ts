import { IPoolsidePlaylist } from "data/constants/playlists.constants";

export interface ISelectProps {
	id: string;
	label?: string;
	options: IPoolsidePlaylist[];
	placeholder?: string;
	onSelectOption?: (option: IPoolsidePlaylist) => void;
}

export interface ISelectState {
	isOpen: boolean;
}
