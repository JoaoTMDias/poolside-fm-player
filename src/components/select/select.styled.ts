import { rem } from "polished";
import styled from "styled-components";

export const SelectWrapper = styled.div`
	--color-select-button: var(--color-primary);
	--color-select-button-text: var(--body-font-color);
	--color-select-button-arrow: var(--body-font-color);
	--color-select-button-border: var(--color-primary);
	--color-select-button-hover: var(--color-primary-dark);
	--color-select-button-text-hover: var(--body-font-color);
	--color-select-button-focus: var(--color-primary-dark);

	html[data-theme="terminal"] && {
		--color-select-button: var(--body-background);
		--color-select-button-text: var(--color-primary);
		--color-select-button-arrow: var(--color-select-button-text);
		--color-select-button-border: var(--color-primary);
		--color-select-button-hover: var(--color-button-hover);
		--color-select-button-text-hover: var(--body-background);
		--color-select-button-focus: var(--color-button-hover);
	}

	width: 100%;
	height: auto;
	display: flex;
	flex-direction: column;
	justify-content: flex-start;
	align-content: center;

	.select-input__container {
		width: 100%;
		display: flex;
		flex-direction: column;
		justify-content: flex-start;
		align-content: center;
		height: calc(var(--global-margin) * 2.5);
		position: relative;
	}

	.select-input__button {
		height: 100%;
		display: flex;
		flex-direction: row;
		justify-content: space-between;
		align-items: center;

		font-family: var(--body-font-family);
		font-size: var(--body-font-size);

		text-transform: capitalize;
		text-shadow: none;
		color: var(--color-black);

		-webkit-appearance: none;
		-moz-appearance: none;
		appearance: none;

		border: 1px solid var(--color-select-button-border);
		background-color: var(--color-select-button);

		cursor: context-menu;

		box-shadow: 0 1px 0 var(--color-black);

		&:hover,
		&:focus,
		&:active {
			background-color: var(--color-select-button-hover);

			.select-input {
				&__label,
				&__value {
					color: var(--color-select-button-text-hover);
				}
			}
		}

		&:focus {
			outline-color: var(--color-select-button-border);
			outline-style: dotted;
			outline-width: 1px;
		}
	}

	.select-input {
		&__left {
			display: flex;
			flex-direction: row;
			justify-content: flex-start;
			align-items: center;
			width: 100%;
		}

		&__arrow {
			display: flex;
			flex-direction: row;
			align-self: flex-end;
			transform: rotate(0deg);
			transition: all 64ms;
			width: ${rem("12px")};
			height: ${rem("18px")};

			path {
				fill: var(--color-select-button-arrow);
			}
		}

		&__label,
		&__value {
			color: var(--color-select-button-text);
		}

		&__label {
			font-family: var(--heading-font-family);
		}

		&__value {
			margin-left: 0.5rem;
			margin-bottom: 0;
			transform: translateY(1px);
		}
	}

	&.is-open {
		.select-input__arrow {
			transform: rotate(180deg);
			transition: all 64ms;
		}
	}
`;

export const SelectOptionsList = styled.ul`
	--color-select-option: var(--color-white);
	--color-select-option-selected: var(--color-gray8);
	--color-select-option-selected-text: var(--color-white);
	--color-select-option-border: var(--color-black);
	--color-select-option-hover: var(--color-primary-dark);
	--color-select-option-focus: var(--color-primary-dark);
	--color-select-option-text: var(--body-font-color);
	--color-select-option-text-hover: var(--body-font-color);
	--color-select-option-new-text: var(--color-black);
	--color-select-option-new-text-background: var(--color-primary);
	--color-select-option-new-text-hover: var(--color-white);

	html[data-theme="terminal"] && {
		--color-select-option: var(--body-background);
		--color-select-option-border: var(--body-background);
		--color-select-option-hover: var(--color-primary);
		--color-select-option-text: var(--color-primary);
		--color-select-option-text-hover: var(--body-background);
		--color-select-option-selected: var(--color-primary);
		--color-select-option-selected-text: var(--body-background);
		--color-select-option-new-text: var(--color-primary);
		--color-select-option-new-text-background: var(--body-background);
		--color-select-option-new-text-hover: var(--body-background);
	}

	display: none;
	box-shadow: 0 2px 10px 0 rgba(0, 0, 0, 0.1);

	border: 1px solid var(--color-select-option-border);
	background-color: var(--color-select-option);

	padding: 0;
	margin: 0;

	max-height: ${rem("148px")};
	overflow-x: hidden;
	overflow-y: scroll;
	-webkit-overflow-scrolling: "smooth";

	&.is-open {
		display: flex;
		flex-direction: column;
		justify-content: flex-start;
		align-items: flex-start;
		position: absolute;
		width: 100%;
		top: calc(var(--global-margin) * 2.5);
		z-index: 100;
	}

	.select-input__option {
		width: 100%;
		display: flex;
		flex-direction: row;
		justify-content: flex-start;
		box-sizing: border-box;
		padding: ${rem("8px")};
		cursor: pointer;
		box-shadow: 0px 1px 0px var(--color-select-option-border);

		&:last-child {
			box-shadow: none;
		}

		&:hover,
		&:focus {
			background-color: var(--color-select-option-hover);
		}

		&[aria-selected="true"] {
			background-color: var(--color-select-option-selected);
			color: var(--color-select-option-selected-text);

			.ui-label {
				outline-color: var(--color-select-option-selected-text);
				outline-style: dotted;
				outline-width: 1px;
			}

			&:hover,
			&:focus {
				background-color: var(--color-select-option-selected);

				.ui-label {
					color: var(--color-select-option-text-hover);
					outline-color: var(--color-select-option-selected-text);
					outline-style: dotted;
					outline-width: 1px;
				}
			}

			.select-input__option__label {
				outline: none;

				&:hover,
				&:focus {
					.ui-label {
						color: var(--color-select-option-text-hover);
						outline-color: var(--color-select-option-selected-text);
						outline-style: dotted;
						outline-width: 1px;
					}
				}
			}
		}
	}

	.select-input__option__label {
		width: 100%;
		height: 100%;
		-webkit-appearance: none;
		border: none;
		background-color: initial;
		color: unset;
		margin: 0;
		padding: 0;
		display: flex;
		flex-direction: row;
		justify-content: flex-start;
		align-items: center;
		cursor: pointer;

		.ui-label {
			width: 100%;
			height: 1.25rem;
			display: flex;
			flex-direction: row;
			justify-content: space-between;
			align-items: center;
		}

		.is-new {
			width: auto;
			background-color: var(--color-select-option-new-text-background);
			color: var(--color-select-option-new-text);
			border-radius: 1rem;
			padding: 0.125rem 0.5rem;
			font-family: var(--heading-font-family);
			text-transform: uppercase;
			letter-spacing: 1px;
			font-size: 10px;
			animation-duration: 1000ms;
			animation-timing-function: step-start;
			animation-delay: 0s;
			animation-iteration-count: infinite;
			animation-direction: normal;
			animation-fill-mode: none;
			animation-play-state: running;
			animation-name: blink;
		}

		&:hover,
		&:focus {
			outline: none;

			.ui-label {
				width: 100%;
				outline-color: var(--color-black);
				outline-style: dotted;
				outline-width: 1px;
			}

			.is-new {
				background-color: var(--color-select-option-new-text-hover);
			}
		}
	}
`;
