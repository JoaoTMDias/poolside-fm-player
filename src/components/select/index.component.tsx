import * as React from "react";
import SelectInputOptions from "./select-input-option";
import { ISelectOption, ISelectProps, ISelectState } from "./select.interface";
import * as S from "./select.styled";

export const defaultProps = {
	id: "select-channel",
	label: "Channel",
	placeholder: "Music channel",
};

/**
 * @class Select
 * @extends {React.Component<ISelectProps, ISelectState>}
 */
class Select extends React.Component<ISelectProps, ISelectState> {
	static defaultProps = defaultProps;

	private selectInputButton: React.RefObject<HTMLButtonElement>;

	constructor(props: ISelectProps) {
		super(props);

		this.state = {
			isOpen: false,
			selectButton: {
				id: "poolside-fm",
				label: "Poolside FM (default)",
				value: "poolside-fm",
			},
		};

		// Bindings
		this.onClickOnSelectButton = this.onClickOnSelectButton.bind(this);

		// Refs
		this.selectInputButton = React.createRef<HTMLButtonElement>();
	}

	/**
	 * Changes the isOpen state to the inverse of what it was
	 *
	 * @returns {void}
	 */
	onClickOnSelectButton() {
		const { isOpen } = this.state;

		this.setState({
			isOpen: !isOpen,
		});
	}

	/**
	 * Handles the onChange event on the list.
	 *
	 * @param {ISelectOption} option
	 * @returns {void}
	 */
	onChangeOptionFromList(
		option: ISelectOption | null,
		event: React.ChangeEvent<HTMLInputElement> | React.KeyboardEvent<HTMLLabelElement>,
	) {
		event.preventDefault();
		const { onSelectOption } = this.props;

		if (option) {
			this.setState(
				{
					selectButton: option,
				},
				() => {
					if (onSelectOption) {
						onSelectOption(option);
					}
				},
			);
		} else {
			event.stopPropagation();
		}

		this.setState(
			{
				isOpen: false,
			},
			() => {
				if (this.selectInputButton.current) {
					this.selectInputButton.current.focus();
				}
			},
		);
	}

	public render() {
		const { id, label, options } = this.props;
		const { isOpen, selectButton } = this.state;

		return (
			<S.SelectWrapper id={id} data-testid="component-select" className={`select-input ${isOpen ? "is-open" : ""}`}>
				<div className="select-input__container">
					<button
						ref={this.selectInputButton}
						key={selectButton.id}
						type="button"
						data-testid="component-select-button"
						id={`${id}-button`}
						aria-haspopup="listbox"
						aria-labelledby={`${id}-label ${id}-title`}
						aria-expanded={isOpen ? "true" : "false"}
						className={`row select-input__button ${isOpen ? "select-input__button--is-open" : ""}`}
						onClick={this.onClickOnSelectButton}
					>
						<span id={`${id}-label`} className="select-input__label">
							{label}
						</span>
						<span id={`${id}-title`} className="select-input__value as-paragraph">
							{selectButton.label}
						</span>
					</button>
					{isOpen && options && (
						<SelectInputOptions
							id={id}
							activeElement={selectButton}
							isOpen={isOpen}
							options={options}
							onChangeOptionFromList={(
								option: ISelectOption | null,
								event: React.ChangeEvent<HTMLInputElement> | React.KeyboardEvent<HTMLLabelElement>,
							) => this.onChangeOptionFromList(option, event)}
						/>
					)}
				</div>
			</S.SelectWrapper>
		);
	}
}

export default Select;
