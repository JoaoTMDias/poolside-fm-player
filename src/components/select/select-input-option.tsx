// Libraries
import * as React from "react";
import { KEY_CODES, holdOn } from "helpers";
import { ISelectOption } from "./select.interface";
import * as S from "./select.styled";

// Interface
interface ISelectInputOptionsProps {
	id: string;
	activeElement: ISelectOption;
	isOpen: boolean;
	options: ISelectOption[];
	onChangeOptionFromList(
		index: number | null,
		event: React.ChangeEvent<HTMLInputElement> | React.KeyboardEvent<HTMLLabelElement>,
	): void;
}

interface ISelectInputOptionsState {
	optionItems: HTMLElement[] | null;
	firstIndex: number;
	currentIndex: number;
	lastIndex: number;
}

const SelectInputOptionLabel: React.FunctionComponent<{
	label: string;
}> = ({ label }) => {
	return <span className="ui-label">{label}</span>;
};

class SelectInputOptions extends React.PureComponent<ISelectInputOptionsProps, ISelectInputOptionsState> {
	constructor(props: ISelectInputOptionsProps) {
		super(props);

		this.state = {
			optionItems: null,
			firstIndex: 0,
			currentIndex: 0,
			lastIndex: 0,
		};
	}

	componentDidMount() {
		this.getAllElementsOfList();
	}

	/**
	 * @description Queries the DOM for all the select input options
	 * @author João Dias
	 * @date 2019-07-11
	 * @memberof SelectInputOptions
	 */
	async getAllElementsOfList() {
		const { options } = this.props;
		const { firstIndex } = this.state;

		if (options && options.length > firstIndex) {
			await holdOn(120);

			const allOptions: HTMLElement[] = Array.from(document.querySelectorAll(".select-input__option"));

			if (allOptions && allOptions.length > 0) {
				let currentIndex = firstIndex;
				const first = allOptions[firstIndex];
				const lastIndex = allOptions.length - 1;

				let focusableElement: HTMLElement | null = null;
				const currentActiveElementIndex =
					allOptions && allOptions.findIndex(option => option.classList.contains("is-selected"));

				if (currentActiveElementIndex && currentActiveElementIndex !== -1) {
					currentIndex = currentActiveElementIndex;
					focusableElement = allOptions[currentIndex];
				} else if (first) {
					focusableElement = first;
				}

				this.setState(
					{
						optionItems: allOptions,
						lastIndex,
						currentIndex,
					},
					() => {
						if (focusableElement) {
							this.setFocusOnOption(focusableElement);
						}
					},
				);
			}
		}
	}

	/**
	 * @description Focus the label on a specific option.
	 * This is visible by the focus ring styling
	 * @author João Dias
	 * @date 2019-07-11
	 * @param {HTMLElement} element
	 * @memberof SelectInputOptions
	 */
	setFocusOnOption(element: HTMLElement) {
		const focusableLabel: HTMLElement | null = element.querySelector(".select-input__option__label");

		if (focusableLabel) {
			focusableLabel.focus();
		}
	}

	/**
	 * @description Creates a roving tab index scheme
	 * @author João Dias
	 * @date 2019-07-11
	 * @param {number} keycode
	 * @returns
	 * @memberof SelectInputOptions
	 */
	handleRovingTabIndex(event: React.KeyboardEvent<HTMLLabelElement>): void {
		event.preventDefault();

		const { currentIndex, firstIndex, optionItems, lastIndex } = this.state;
		let nextIndex = currentIndex;

		if (optionItems) {
			if (event.keyCode === KEY_CODES.ARROW_UP) {
				nextIndex = currentIndex - 1;
			} else if (event.keyCode === KEY_CODES.ARROW_DOWN || event.keyCode === KEY_CODES.TAB) {
				if (event.shiftKey) {
					nextIndex = currentIndex - 1;
				} else {
					nextIndex = currentIndex + 1;
				}
			}

			if (nextIndex > lastIndex) {
				nextIndex = firstIndex;
			} else if (nextIndex < firstIndex) {
				nextIndex = lastIndex;
			}

			this.setFocusOnOption(optionItems[nextIndex]);
			this.setState({
				currentIndex: nextIndex,
			});
		}
	}

	/**
	 * @description Handles the key down presses whilst on the select input
	 * @author João Dias
	 * @date 2019-07-11
	 * @param {number} index
	 * @param {React.KeyboardEvent<HTMLLabelElement>} event
	 * @memberof SelectInputOptions
	 */
	handleOnKeyDownPress(index: number, event: React.KeyboardEvent<HTMLLabelElement>) {
		const { onChangeOptionFromList } = this.props;

		switch (event.keyCode) {
			case KEY_CODES.SPACE:
			case KEY_CODES.ENTER:
				onChangeOptionFromList(index, event);
				break;

			case KEY_CODES.ESC:
				onChangeOptionFromList(null, event);
				break;

			case KEY_CODES.TAB:
			case KEY_CODES.ARROW_DOWN:
			case KEY_CODES.ARROW_UP:
			case KEY_CODES.SHIFT && KEY_CODES.TAB:
				this.handleRovingTabIndex(event);
				break;

			default:
				break;
		}
	}

	/**
	 * @description Constroi uma lista de opções
	 * @author João Dias
	 * @date 2019-06-28
	 * @param {ISelectOption[]} options
	 * @returns
	 * @memberof SelectInputOptions
	 */
	renderListOfOptions() {
		const { id, options, onChangeOptionFromList, activeElement } = this.props;

		const list = options.map((option: ISelectOption, index: number) => {
			const key = `${option.id}-id-${index}`;
			const isSelected = !!(activeElement.id === option.id);

			const element = (
				<li
					key={key}
					id={key}
					role="option"
					className={`select-input__option ${isSelected && "is-selected"}`}
					aria-selected={isSelected}
				>
					<label
						className="select-input__option__label"
						htmlFor={`${option.id}-input`}
						tabIndex={0}
						onKeyDown={(event: React.KeyboardEvent<HTMLLabelElement>) => {
							this.handleOnKeyDownPress(index, event);
						}}
					>
						<input
							id={`${option.id}-input`}
							name={`${id}-select-input-options`}
							value={option.id}
							type="radio"
							className="sr-only"
							checked={isSelected}
							onChange={(event: React.ChangeEvent<HTMLInputElement>) => onChangeOptionFromList(index, event)}
							tabIndex={-1}
						/>

						<SelectInputOptionLabel label={`${option.label}`} />
					</label>
				</li>
			);

			return element;
		});

		return list;
	}

	render() {
		const { activeElement, id, isOpen } = this.props;

		return (
			<S.SelectOptionsList
				id={`${id}-list`}
				aria-labelledby={`${id}-label`}
				aria-activedescendant={activeElement.id}
				data-testid="component-select-list"
				role="listbox"
				className={`${isOpen && "is-open"} custom-scrollbar`}
				tabIndex={-1}
			>
				{this.renderListOfOptions()}
			</S.SelectOptionsList>
		);
	}
}

export default SelectInputOptions;
