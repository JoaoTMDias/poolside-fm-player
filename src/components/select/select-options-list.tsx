// Libraries
import * as React from "react";
import { ISelectOption } from "data/constants";
import { RovingTabIndexProvider } from "helpers/custom-hooks/roving-index";
import * as S from "./select.styled";
import { SelectOption } from "./select-option";
import { useDidMount } from "helpers";

// Interface
interface ISelectInputOptionsListProps {
	id: string;
	activeElement: ISelectOption;
	isOpen: boolean;
	options: ISelectOption[];
	onChangeOptionFromList(index: number | null): void;
}

const SelectInputOptionsList: React.FunctionComponent<ISelectInputOptionsListProps> = (props) => {
	const { activeElement, id, isOpen, options, onChangeOptionFromList } = props;

	useDidMount(() => {
		const element = document.getElementById(activeElement.id);

		if (element) {
			element.focus();
		}
	});

	/**
	 * @description Constroi uma lista de opções
	 * @author João Dias
	 * @date 2019-06-28
	 * @param {ISelectOption[]} options
	 * @returns
	 * @memberof SelectInputOptionsList
	 */
	function renderListOfOptions() {
		const list = options.map((option: ISelectOption, index: number) => {
			const key = option.id;

			const element = (
				<SelectOption
					key={key}
					id={key}
					onSelect={() => onChangeOptionFromList(index)}
					option={option}
				/>
			);

			return element;
		});

		return list;
	}

	return (
		<S.SelectOptionsList
			id={`${id}-list`}
			aria-labelledby={`${id}-label`}
			aria-activedescendant={activeElement.id}
			data-testid="component-select-list"
			role="listbox"
			className={`${isOpen && "is-open"} custom-scrollbar`}
			tabIndex={-1}
		>
			<RovingTabIndexProvider>{renderListOfOptions()}</RovingTabIndexProvider>
		</S.SelectOptionsList>
	);
};

export default SelectInputOptionsList;
