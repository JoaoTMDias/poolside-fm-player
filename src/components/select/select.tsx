import * as React from "react";
import { ISelectOption } from "data/constants";
import SelectInputOptions from "./select-option";
import { ISelectProps, ISelectState } from "./select.interface";
import * as S from "./select.styled";

export const defaultProps = {
	id: "select-channel",
	label: "Channel",
	placeholder: "Music channel",
	type: "playlist",
};

/**
 * @class Select
 * @extends {React.Component<ISelectProps, ISelectState>}
 */
class Select extends React.Component<ISelectProps, ISelectState> {
	private selectInputButton: React.RefObject<HTMLButtonElement>;

	static defaultProps = defaultProps;

	constructor(props: ISelectProps) {
		super(props);

		this.state = {
			isOpen: false,
		};

		// Bindings
		this.onClickOnSelect = this.onClickOnSelect.bind(this);

		// Refs
		this.selectInputButton = React.createRef<HTMLButtonElement>();
	}

	/**
	 * Changes the isOpen state to the inverse of what it was
	 *
	 * @returns {void}
	 */
	onClickOnSelect() {
		const { isOpen } = this.state;

		this.setState({
			isOpen: !isOpen,
		});
	}

	/**
	 * Handles the onChange event on the list.
	 *
	 * @param {(number | null)} index
	 * @param {(React.ChangeEvent<HTMLInputElement> | React.KeyboardEvent<HTMLLabelElement>)} event
	 * @memberof Select
	 */
	onChangeOptionFromList(
		index: number | null,
		event: React.ChangeEvent<HTMLInputElement> | React.KeyboardEvent<HTMLLabelElement>,
	) {
		event.preventDefault();
		const { onChange } = this.props;
		const hasIndex = index && index >= 0;

		if (onChange && index && hasIndex) {
			onChange(index);
		} else {
			event.stopPropagation();
		}

		this.setState(
			{
				isOpen: false,
			},
			() => {
				if (this.selectInputButton.current) {
					this.selectInputButton.current.focus();
				}
			},
		);
	}

	/**
	 * Renders the select component
	 *
	 * @param {(ISelectOption[])} options
	 * @param {number} currentIndex
	 * @param {string} id
	 * @param {boolean} isOpen
	 * @param {(string | undefined)} label
	 * @returns
	 * @memberof Select
	 */
	renderSelectComponent(
		options: ISelectOption[],
		currentIndex: number,
		id: string,
		isOpen: boolean,
		label: string | undefined,
	) {
		const selected = options[currentIndex];
		return (
			<div className="select-input__container">
				<button
					ref={this.selectInputButton}
					key={selected.id}
					type="button"
					data-testid="component-select-button"
					id={`${id}-button`}
					aria-haspopup="listbox"
					aria-labelledby={`${id}-label ${id}-title`}
					aria-expanded={isOpen ? "true" : "false"}
					className={`row select-input__button ${isOpen ? "select-input__button--is-open" : ""}`}
					onClick={this.onClickOnSelect}
				>
					<span id={`${id}-label`} className="select-input__label">
						{label}
					</span>
					<span id={`${id}-title`} className="select-input__value as-paragraph">
						{selected.label}
					</span>
				</button>
				{isOpen && options && (
					<SelectInputOptions
						id={id}
						activeElement={selected}
						isOpen={isOpen}
						options={options}
						onChangeOptionFromList={(
							index: number | null,
							event: React.ChangeEvent<HTMLInputElement> | React.KeyboardEvent<HTMLLabelElement>,
						) => this.onChangeOptionFromList(index, event)}
					/>
				)}
			</div>
		);
	}

	public render() {
		const { id, label, options, currentIndex } = this.props;
		const { isOpen } = this.state;

		return (
			<S.SelectWrapper id={id} data-testid="component-select" className={`select-input ${isOpen ? "is-open" : ""}`}>
				{this.renderSelectComponent(options, currentIndex, id, isOpen, label)}
			</S.SelectWrapper>
		);
	}
}

export default Select;
