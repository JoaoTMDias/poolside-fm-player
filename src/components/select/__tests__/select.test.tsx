import React from "react";
import { render, cleanup, fireEvent } from "@testing-library/react"

import Select from "components/select/select";
import { ISelectProps } from "../select.interface";

const initialProps: ISelectProps = {
	id: "component-select",
	currentIndex: 0,
	options: [
		{
			id: "poolside-fm",
			label: "Poolside FM (default)",
			value: "poolside-fm",
		},
		{
			id: "indie-summer",
			label: "Indie Summer",
			value: "indie-summer",
		},
		{
			id: "hangover-club",
			label: "Hangover Club",
			value: "hangover-club",
		},
		{
			id: "tokyo-disco",
			label: "Tokyo Disco",
			value: "tokyo-disco",
		},
		{
			id: "friday-nite-heat",
			label: "Friday Nite Heat",
			value: "friday-nite-heat",
		},
	],
};

afterEach(cleanup);

describe("<Select />", () => {
	it("should render a select component", () => {
		const component = render(<Select {...initialProps} />);

		expect(component).toMatchSnapshot();
	});

	it("should be closed by default", async () => {
		const { queryByTestId } = render(<Select {...initialProps} />);
		const selectOptionsList = await queryByTestId("component-select-list");
		expect(selectOptionsList).toBeNull();
	});

	it("should display a list on click on the button", async () => {
		const { getByTestId, getAllByTestId } = render(<Select {...initialProps} />);
		const selectButton = await getByTestId("component-select-button");

		fireEvent.click(selectButton);

		const selectOptionsList = await getAllByTestId("component-select-list");
		expect(selectOptionsList.length).toBe(1);
	});

	it("should display a default label on the button", async () => {
		const { getByTestId } = render(<Select {...initialProps} />);
		const selectButton = await getByTestId('component-select-button');
		const selectButtonValue = await selectButton.querySelector(".select-input__value");

		expect(selectButtonValue?.textContent).toBe(initialProps.options[0].label);
	});
});
