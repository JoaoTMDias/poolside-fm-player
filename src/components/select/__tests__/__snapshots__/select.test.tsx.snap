// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> should render a select component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-AxjAm goVOyD select-input "
        data-testid="component-select"
        id="component-select"
      >
        <div
          class="select-input__container"
        >
          <button
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-labelledby="component-select-label component-select-title"
            class="row select-input__button "
            data-testid="component-select-button"
            id="component-select-button"
            type="button"
          >
            <div
              class="select-input__left"
            >
              <span
                class="select-input__label"
                id="component-select-label"
              >
                Channel
              </span>
              <span
                class="select-input__value as-paragraph"
                id="component-select-title"
              >
                Poolside FM (default)
              </span>
            </div>
            <svg
              class="select-input__arrow"
              fill="none"
              height="13"
              viewBox="0 0 12 13"
              width="12"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M6 10.35l-4.157-5.4h8.314L6 10.35z"
                fill="var(--color-select-button-arrow)"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-AxjAm goVOyD select-input "
      data-testid="component-select"
      id="component-select"
    >
      <div
        class="select-input__container"
      >
        <button
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-labelledby="component-select-label component-select-title"
          class="row select-input__button "
          data-testid="component-select-button"
          id="component-select-button"
          type="button"
        >
          <div
            class="select-input__left"
          >
            <span
              class="select-input__label"
              id="component-select-label"
            >
              Channel
            </span>
            <span
              class="select-input__value as-paragraph"
              id="component-select-title"
            >
              Poolside FM (default)
            </span>
          </div>
          <svg
            class="select-input__arrow"
            fill="none"
            height="13"
            viewBox="0 0 12 13"
            width="12"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6 10.35l-4.157-5.4h8.314L6 10.35z"
              fill="var(--color-select-button-arrow)"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
